x/upgrade/keeper/keeper.go

```diff
@@ -177,107 +197,152 @@ func (k Keeper) GetModuleVersions(ctx sdk.Context) []*types.ModuleVersion {
 }
 
-// GetUpgradedClient gets the expected upgraded client for the next version of this chain
-func (k Keeper) GetUpgradedClient(ctx sdk.Context, height int64) ([]byte, bool) {
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.UpgradedClientKey(height))
-	if len(bz) == 0 {
-		return nil, false
+// GetUpgradedClient gets the expected upgraded client for the next version of this chain. If not found it returns
+// ErrNoUpgradedClientFound, but other errors may be returned if there is an error reading from the store.
+func (k Keeper) GetUpgradedClient(ctx context.Context, height int64) ([]byte, error) {
+	store := k.storeService.OpenKVStore(ctx)
+	bz, err := store.Get(types.UpgradedClientKey(height))
+	if err != nil {
+		return nil, err
+	}
+
+	if bz == nil {
+		return nil, types.ErrNoUpgradedClientFound
 	}
 
-	return bz, true
+	return bz, nil
 }
 
 // SetUpgradedConsensusState sets the expected upgraded consensus state for the next version of this chain
 // using the last height committed on this chain.
-func (k Keeper) SetUpgradedConsensusState(ctx sdk.Context, planHeight int64, bz []byte) error {
-	store := ctx.KVStore(k.storeKey)
-	store.Set(types.UpgradedConsStateKey(planHeight), bz)
-	return nil
+func (k Keeper) SetUpgradedConsensusState(ctx context.Context, planHeight int64, bz []byte) error {
+	store := k.storeService.OpenKVStore(ctx)
+	return store.Set(types.UpgradedConsStateKey(planHeight), bz)
 }
 
-// GetUpgradedConsensusState gets the expected upgraded consensus state for the next version of this chain
-func (k Keeper) GetUpgradedConsensusState(ctx sdk.Context, lastHeight int64) ([]byte, bool) {
-	store := ctx.KVStore(k.storeKey)
-	bz := store.Get(types.UpgradedConsStateKey(lastHeight))
-	if len(bz) == 0 {
-		return nil, false
+// GetUpgradedConsensusState gets the expected upgraded consensus state for the next version of this chain. If not found
+// it returns ErrNoUpgradedConsensusStateFound, but other errors may be returned if there is an error reading from the store.
+func (k Keeper) GetUpgradedConsensusState(ctx context.Context, lastHeight int64) ([]byte, error) {
+	store := k.storeService.OpenKVStore(ctx)
+	bz, err := store.Get(types.UpgradedConsStateKey(lastHeight))
+	if err != nil {
+		return nil, err
 	}
 
-	return bz, true
+	if bz == nil {
+		return nil, types.ErrNoUpgradedConsensusStateFound
+	}
+
+	return bz, nil
 }
 
 // GetLastCompletedUpgrade returns the last applied upgrade name and height.
-func (k Keeper) GetLastCompletedUpgrade(ctx sdk.Context) (string, int64) {
-	iter := storetypes.KVStoreReversePrefixIterator(ctx.KVStore(k.storeKey), []byte{types.DoneByte})
-	defer iter.Close()
+func (k Keeper) GetLastCompletedUpgrade(ctx context.Context) (string, int64, error) {
+	store := k.storeService.OpenKVStore(ctx)
+	prefix := []byte{types.DoneByte}
+	it, err := store.ReverseIterator(prefix, storetypes.PrefixEndBytes(prefix))
+	if err != nil {
+		return "", 0, err
```

Comment:
if we return here do we need to close the iterator?

Reply:
I'll check all of these. I'm going to add a deferred close, although it will ignore the error tho

Reply:
Its best to close the iterator at error return statement too, just to ensure that there is no resource leakage.